### ISSUE TITLE
Forwarded emails not getting fetched from mailbox

### ISSUE THREAD

**Issue by slawn** [uvdesk/community-skeleton#376]
When you have a mailbox and emails are forwarded to it - the forwarded emails are not getting fetched to the ticketing system.

To give an example 
- mailbox is test@test.com
- email from client is sent to forward@test.com which is then forwarded to test@test.com

Those emails are not fetched to the ticketing system. I guess this might have something to do with the fact that when email is forwarded header TO stays the same - so in example above TO header would be still forward@test.com.

Bottom line - those tickets are not created. The assumption would be that any emails that sit in mailbox should be converted into tickets which is not the case currently. This is easy to troubleshoot - try to setup forward email to any existing mailbox and mails won't be fetched.

Using latest version installed with composer and PHP 7.2.

**Comment by slawn** [comment]
After further investigation I have found there are 2 major / fundamental issues here - which I think requires immediate attention.

1. Emails that are forwarded to mailbox - will never get fetched
2. If client is sending ticket to a mailbox but there are multiple TO addresses and mailbox email is not the first one - email will never get fetched. To give an example if a user will send an email to someaddress@test.com,mailboxaddress@test.com and mailbox email will be the second one - email will never get fetched. Clients may with to send an email to few people - happens to us all the time that they also want to notify manager about something and they simply send to multiple recipients.

The issue is how you are determining the recipient of the email. In this file:

vendor/uvdesk/mailbox-component/Services/MailboxService.php

in processMail method - you set $addresses['to'] - which is the issue. The problem with forwarded emails is that TO is the email address where it was sent originally - which will then fail your checks. It's similar when you are sending to multiple recipients. To explain - once the $addresses['to'] is set - then it's using createTicket method in vendor/uvdesk/core-framework/Services/TicketService.php - over there you are setting the mailbox email using this:

$params['mailboxEmail'] = current($params['replyTo']); 

which means you are picking the first email from the $params['replyTo'] array of emails. Now in case of the forwarded email - it will be the email where the email was sent originally before being forwarded and in case of multiple recipients you only pick the first one - so if the mailbox email was second it will fail. This is because before the email is inserted into DB you do a check for mailbox - if the email is there - you do it in createTicketBase in  vendor/uvdesk/core-framework/Services/TicketService.php.  

$mailbox = $this->container->get('uvdesk.mailbox')->getMailboxByEmail($ticketData['mailboxEmail']);

If this will fail email is not fetched.

I've fixed it on my installation by doing one simple assumption - instead of trying to get the recipient from email - I get it from the inbox settings. To explain - in refreshMailbox method in vendor/uvdesk/mailbox-component/Console/RefreshMailboxCommand.php I added 3rd parameter to 

$this->pushMessage($message, $useSecureConnection, $server_username); 

I'm passing server username so mailbox email. Then in pushMessage method in vendor/uvdesk/mailbox-component/Console/RefreshMailboxCommand.php I'm posting this email along with the raw email like that:

curl_setopt($curlHandler, CURLOPT_POSTFIELDS, http_build_query(['email' => $message, 'email_to' => $server_username]));

and then in your processMail method in vendor/uvdesk/mailbox-component/Services/MailboxService.php I simply use that email as the TO email. It's a safe assumption that when you are fetching emails from certain mailbox - that the TO should be the mailbox email. This was both issues are resolved - all forwarded emails work and if user sends email to multiple recipients at once and mailbox email is not the first one - it still works. 

Happy to share the code if you want to have a look and see what I mean. For me it's a fundamental issue as both forwards are fairly standard and tickets should also be picked up if users send to multiple recipients.

Thanks

**Comment by papnoisanjeev** [comment]
@slawn 

Thanks you your suggestions and updates  we will test these changes and update with project If everything goes good.
Still you found any fix for any issue somewhere you can create a pull request (with your updated code) for the same.

**Comment by papnoisanjeev** [comment]
@slawn 

We have updated code for forward messages.

Locate path in your project:
**vendor/uvdesk/mailbox-component/Services/MailboxService.php**

Open **MailBoxService.php** and replace its code from [here](https://github.com/uvdesk/mailbox-component/blob/master/Services/MailboxService.php).

**Comment by PeopleInside** [comment]
For me this update stop ticket to be opened by email https://github.com/uvdesk/mailbox-component/issues/67

**Comment by slawn** [comment]
> @slawn
> 
> We have updated code for forward messages.
> 
> Locate path in your project:
> **vendor/uvdesk/mailbox-component/Services/MailboxService.php**
> 
> Open **MailBoxService.php** and replace its code from [here](https://github.com/uvdesk/mailbox-component/blob/master/Services/MailboxService.php).

For me that doesn't really work - I have emails forwarded with office365 and with cPanel - none of them add this header (simply not there). I still revert to my solution of assuming TO from the inbox that is checked - all emails in certain inbox have the same TO for me - the most logical solution. Also this fix doesn't resolve the issue with multiple TO where support email is not the first one - so good attempt but not a successful one. Thanks for trying though.

**Comment by papnoisanjeev** [comment]
@slawn 

Are you not able to covert ticket after replacing the **MailBoxService.php** ?

**Comment by papnoisanjeev** [comment]
@slawn 

We have recently made a new [release ](https://github.com/uvdesk/community-skeleton/releases/tag/v1.0.13
) of project.

You can check now.

**Comment by papnoisanjeev** [comment]
@slawn 

We are closing the issue as it is fixed from our side.
If still exist you can reopen a new again.

**Comment by MrBryan** [comment]
> and then in your processMail method in vendor/uvdesk/mailbox-component/Services/MailboxService.php I simply use that email as the TO email. It's a safe assumption that when you are fetching emails from certain mailbox - that the TO should be 

Can you please share the changes you made to this file -- I can't get the forwarded mail presently either

**Comment by MrBryan** [comment]
@slawn

**Comment by papnoisanjeev** [comment]
@MrBryan 

Which version of **UVdesk opensource** you are using?

Also replace [MailboxService.php](https://github.com/uvdesk/community-skeleton/issues/376#issuecomment-804158851) if using old version.

**Comment by slawn** [comment]
> > and then in your processMail method in vendor/uvdesk/mailbox-component/Services/MailboxService.php I simply use that email as the TO email. It's a safe assumption that when you are fetching emails from certain mailbox - that the TO should be
> 
> Can you please share the changes you made to this file -- I can't get the forwarded mail presently either

Here you can see my changes: https://github.com/slawn/mailbox-component/commit/1617a04b69fd06faf9bb2c1c87b05199998b90f1

Just don't grab the whole file if you are using new version as I haven't updated that since I did it. You would have to copy the changes one by one.
